File: python/stacks/stacks_2.py
class StackSet:
    def __init__(self, stack_capacity=None):
        self.stacks = []
        self.index = None
        self.stack_capacity = stack_capacity

    def __repr__(self):
        stacks = ""

        for i, s in enumerate(self.stacks):
            stacks += f'{i}: {s}\n'

        return stacks

    def pop(self):
        if self.index == None:
            raise Exception('stack set is empty')

        # find a non-empty stack
        # necessary to support popAt(i)
        while not self.stacks[self.index].top:
            # Python list built in pop
            self.stacks.pop()
            self.index -= 1

            if self.index == -1:
                self.index = None
                raise Exception('stack set is empty')

        return self.stacks[self.index].pop()

    def push(self, data):
        if self.index == None:
            self.index = 0
            self.stacks.append(Stack())
            self.push(data)
        elif self.stacks[self.index].length == self.stack_capacity:
            self.index += 1
            self.stacks.append(Stack())
            self.push(data)
        else:
            self.stacks[self.index].push(data)

    def popAt(self, i):
        if i >= len(self.stacks):
            raise IndexError

        if not self.stacks[i].top:
            raise Exception('stack is empty')

        return self.stacks[i].pop()


class Stack:
    def __init__(self, top=None):
        self.top = top
        self.length = 0

    def __repr__(self):
        frames = []
        top = self.top

        while top:
            frames.append(str(top))
            top = top.next

        if frames:
            return f'{" -> ".join(["top"] + frames + ["bottom"])}'

        return 'empty stack'

    def push(self, data):
        frame = Frame(data, self.top)
        self.top = frame
        self.length += 1

    def pop(self):
        if not self.top:
            raise Exception('stack is empty')

        frame = self.top
        self.top = self.top.next
        self.length -= 1

        return frame.data


class Frame:
    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next

    def __repr__(self):
        return f'[ {self.data} ]'


ss = StackSet(5)
ss.push(1)
ss.push(2)
ss.push(3)
ss.push(4)
ss.push(5)
ss.push(1)
ss.push(2)
ss.push(3)
ss.push(4)
ss.push(5)
ss.push(1)
ss.push(2)
ss.push(3)
ss.push(4)
ss.push(5)

print(ss)

print(ss.popAt(1))
print(ss.popAt(1))
print(ss.popAt(1))
print(ss.popAt(1))
print(ss.popAt(1))

print(ss)

print(ss.pop())
print(ss.pop())
print(ss.pop())
print(ss.pop())
print(ss.pop())

print(ss)

print(ss.pop())
print(ss)

File: go/pkg/stacks/stacks_2.go
package stacks

import (
	"fmt"
	"strings"
)

// StackSet is LIFO data structure
// with multiple stacks
// supporting O(1) push, pop, and popAt
type StackSet struct {
	Stacks        []*Stack
	Index         int
	StackCapacity int
}

// ToString provides a string representation for a stack set
func (ss *StackSet) ToString() string {
	stacks := ""

	for i, s := range ss.Stacks {
		stacks += fmt.Sprintf("%d: %s\n", i, s.ToString())
	}

	return stacks
}

// Pop is O(1)
func (ss *StackSet) Pop() int {
	if len(ss.Stacks) == 0 {
		panic("stack set is empty")
	}

	for ss.Stacks[ss.Index].Top == nil {
		ss.Stacks = ss.Stacks[:len(ss.Stacks)-1]
		ss.Index--

		if ss.Index == -1 {
			ss.Index = 0
			panic("stack set is empty")
		}
	}

	return ss.Stacks[ss.Index].Pop()
}

// Push is O(1)
func (ss *StackSet) Push(data int) {
	if len(ss.Stacks) == 0 {
		ss.Stacks = append(ss.Stacks, &Stack{})
		ss.Push(data)
	} else if ss.Stacks[ss.Index].Length == ss.StackCapacity {
		ss.Index++
		ss.Stacks = append(ss.Stacks, &Stack{})
		ss.Push(data)
	} else {
		ss.Stacks[ss.Index].Push(data)
	}
}

// PopAt is O(1)
// this does not move frames between stacks
func (ss *StackSet) PopAt(i int) int {
	if i >= len(ss.Stacks) {
		panic("index out of bounds")
	}

	if ss.Stacks[i].Top == nil {
		panic("stack is empty")
	}

	return ss.Stacks[i].Pop()
}

// Stack is a LIFO data structure
type Stack struct {
	Top    *Frame
	Length int
}

// ToString provides a string representation for a stack
func (s *Stack) ToString() string {
	frames := []string{}
	top := s.Top

	for top != nil {
		frames = append(frames, top.ToString())
		top = top.Next
	}

	if len(frames) > 0 {
		ss := []string{"top"}
		ss = append(ss, frames...)
		ss = append(ss, "bottom")

		return strings.Join(ss, " -> ")
	}

	return "empty stack"
}

// Push O(1)
func (s *Stack) Push(data int) {
	frame := &Frame{Data: data, Next: s.Top}
	s.Top = frame
	s.Length++
}

// Pop O(1)
func (s *Stack) Pop() int {
	if s.Top == nil {
		panic("stack is empty")
	}

	frame := s.Top
	s.Top = s.Top.Next
	s.Length--

	return frame.Data
}

// Frame is a single stack frame
type Frame struct {
	Data int
	Next *Frame
}

// ToString provides a string representation for a stack frame
func (f *Frame) ToString() string {
	return fmt.Sprintf("[ %d ]", f.Data)
}
